### PR TITLE
[JUJU-1758] Developed the bash-based version of nw-cross-model-relations

### PR DESCRIPTION
**Pull request by anvial**
This PR adds bash-based version of CMR tests.
These tests will exercise the following aspects:
  - Ensure a user is able to create an offer of an applications' endpoint
    including:
      - A user is able to consume and relate to the offer
      - Workload data successfully provided
      - The offer appears in the list-offers output
      - The user is able to name the offer
      - The user is able to remove the offer
  - Ensure an admin can grant a user access to an offer
      - The consuming user finds the offer via 'find-offer'

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made

## QA steps

```sh
cd tests
./main.sh -v -p lxd cmr test_offer_consume
```

### CODE REVIEW

File: tests/suites/cmr/offer_consume.sh [hunk 1]
```diff
@@ -0,0 +1,134 @@
+# Exercising the CMR Juju functionality which allows applications to
+# communicate between different models.
+# This test will exercise the following aspects:
+#  - Ensure a user is able to create an offer of an applications' endpoint
```
Comment (wallyworld) [inline tests/suites/cmr/offer_consume.sh:4]: application's

File: tests/suites/cmr/offer_consume.sh [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+#    including:
+#      - A user is able to consume and relate to the offer
+#      - Workload data successfully provided
+#      - The offer appears in the list-offers output
+#      - The user is able to name the offer
+#      - The user is able to remove the offer
```
Comment (wallyworld) [inline tests/suites/cmr/offer_consume.sh:10]: we're not testing all of these things in this particular file though?

File: tests/suites/cmr/offer_consume.sh [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+	wait_for "dummy-sink" "$(idle_condition "dummy-sink")"
+
+	echo "Relate workload in consume model with offer"
+	juju --show-log consume "${BOOTSTRAPPED_JUJU_CTRL_NAME}:admin/model-offer.dummy-offer"
+	juju --show-log relate dummy-sink dummy-offer
+	# wait for relation joined before migrate.
+	wait_for "dummy-offer" '.applications["dummy-sink"] | .relations.source[0]'
+
+	echo "Provide config if offered workload and change the status of consumed offer"
```
Comment (wallyworld) [inline tests/suites/cmr/offer_consume.sh:43]: typo? "if"

File: tests/suites/cmr/offer_consume.sh [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	juju switch "model-offer"
+	juju config dummy-source token=yeah-boi
+	juju switch "model-consume"
+	wait_for "active" '."application-endpoints"["dummy-offer"]."application-status".current'
+
+	echo "Remove offer"
+	# The offer must be removed before model/controller destruction will work.
+	# See discussion under https://bugs.launchpad.net/juju/+bug/1830292.
+	juju switch "model-offer"
+	juju remove-offer "admin/model-offer.dummy-offer" --force -y
```
Comment (wallyworld) [inline tests/suites/cmr/offer_consume.sh:55]: we should test without --force
test that remove fails with relation still present
then remove relation and remove offer

File: tests/suites/cmr/offer_consume.sh [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	juju switch "${offer_controller}:model-offer"
+	juju config dummy-source token=yeah-boi
+	juju switch "controller-consume:model-consume"
+	wait_for "active" '."application-endpoints"["dummy-source"]."application-status".current'
+
+	echo "Remove offer"
+	# The offer must be removed before model/controller destruction will work.
+	# See discussion under https://bugs.launchpad.net/juju/+bug/1830292.
+	juju switch "${offer_controller}:model-offer"
+	juju remove-offer "${offer_controller}:admin/model-offer.dummy-source" --force -y
```
Comment (wallyworld) [inline tests/suites/cmr/offer_consume.sh:111]: same comment as above

File: tests/suites/cmr/task.sh [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	set_verbosity
+
+	echo "==> Checking for dependencies"
+	check_dependencies juju
+
+	file="${TEST_DIR}/test-cmr.log"
+
+	bootstrap "test-cmr" "${file}"
+
+	# Tests that need to be run are added here.
```
Comment (wallyworld) [inline tests/suites/cmr/task.sh:16]: update comment

File: tests/suites/model/migration.sh [hunk 7]
```diff
@@ -401,21 +401,3 @@ test_model_migration_saas_external() {
 		run "run_model_migration_saas_external"
 	)
 }
-
-bootstrap_alt_controller() {
```
Comment (wallyworld) [inline tests/suites/model/migration.sh:405]: is this unused?

**Review by wallyworld — COMMENTED** [review]
a good start

File: tests/suites/model/migration.sh (hunk 7 above)
Comment (anvial) [reply tests/suites/model/migration.sh:405]: no, as this function is used in several tests: `cmr`, and `migration`,  there is a suggestion to move it to the `includes` folder.

File: tests/suites/cmr/offer_consume.sh (hunk 2 above)
Comment (anvial) [reply tests/suites/cmr/offer_consume.sh:10]: I've added `list-offers` and `find-offers` checks. And offer naming, so looks like we now check all these points.

File: tests/suites/cmr/task.sh (hunk 6 above)
Comment (anvial) [reply tests/suites/cmr/task.sh]: I've just removed this comment

**Review by manadart — APPROVED** [review]
We should have a K8s-based CMR test too, but I assume that will be the product of the COS investigation by @basebandit.

**Pull request merged**